nvme.c

```diff
@@ -221,70 +217,82 @@ void *nvme_realloc(void *p, size_t len)
 	return result;
 }
 
-static void *__nvme_alloc_huge(size_t len, bool *huge)
-{
+#define HUGE_MIN 0x80000
+
+struct nvme_mem_huge {
+	size_t len;
+	bool nvme_malloc; /* p has been allocated using nvme_malloc */
 	void *p;
+};
 
-	if (!posix_memalign(&p, getpagesize(), len)) {
-		*huge = false;
-		memset(p, 0, len);
-		return p;
-	}
-	return NULL;
-}
+struct nvme_mem_huge *nvme_alloc_huge(size_t len)
+{
+	struct nvme_mem_huge *mh;
 
-#define HUGE_MIN 0x80000
+	mh = malloc(sizeof(*mh));
+	if (!mh)
+		return NULL;
 
-#ifdef CONFIG_LIBHUGETLBFS
-void nvme_free_huge(void *p, bool huge)
-{
-	if (huge) {
-		if (p)
-			free_hugepage_region(p);
-	} else {
-		free(p);
+	memset(mh, 0, sizeof(*mh));
+
+	mh->len = ROUND_UP(len, 0x1000);
+
+	/*
+	 * For smaller allocation we just use malloc and hope the kernel is able
+	 * to convert to continious memory region.
+	 */
+	if (mh->len < HUGE_MIN) {
+		mh->nvme_malloc = true;
+		mh->p = nvme_alloc(mh->len);
+		return mh;
 	}
-}
 
-void *nvme_alloc_huge(size_t len, bool *huge)
-{
-	void *p;
+	/*
+	 * Larger allocation will almost certainly fail with the malloc
+	 * approach. Instead try pre-allocating memory from the TBH pool.
+	 */
+	mh->p = mmap(NULL, mh->len, PROT_READ | PROT_WRITE,
+		     MAP_ANONYMOUS | MAP_PRIVATE | MAP_HUGETLB, -1, 0);
+	if (mh->p != MAP_FAILED) {
+		return mh;
+	}
 
-	if (len < HUGE_MIN)
-		return __nvme_alloc_huge(len, huge);
+	/*
+	 * And if mmap fails because the pool has no pages available, try to use
+	 * malloc/madvise as fallback in order to fullfil the request. This
+	 * gives the kernel a chance to try to claim some huge pages. This might
+	 * still fail though.
+	 */
+	mh->p = nvme_alloc(mh->len);
```

Comment:
This is going to request 4k, but if I'm reading the man pages correctly, I think we have a higher chance of transparent huge page success if we align to 2MB.